File: it-agents/skills/code-architect/references/code-examples/go/strategy.go
// Strategy: define a family of interchangeable algorithms behind a common
// interface and select one at runtime.
package main

import "fmt"

// Strategy is the common interface for interchangeable algorithms.
type Strategy interface {
	Combine(a, b int) int
}

type sumStrategy struct{}

func (sumStrategy) Combine(a, b int) int { return a + b }

type maxStrategy struct{}

func (maxStrategy) Combine(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// context delegates the work to whichever strategy it holds.
type context struct {
	strategy Strategy
}

func (c *context) Execute(a, b int) int {
	return c.strategy.Combine(a, b)
}

func main() {
	c := &context{strategy: sumStrategy{}}
	fmt.Println("sum:", c.Execute(3, 7))

	c.strategy = maxStrategy{}
	fmt.Println("max:", c.Execute(3, 7))
}

File: it-agents/skills/code-architect/references/code-examples/cpp/Strategy.cpp
// Strategy — a family of interchangeable algorithms behind one interface.
#include <iostream>
#include <memory>
#include <numeric>
#include <vector>

// Strategy interface.
class Reducer {
public:
    virtual ~Reducer() = default;
    virtual int combine(const std::vector<int>& data) const = 0;
};

class SumReducer : public Reducer {
public:
    int combine(const std::vector<int>& data) const override {
        return std::accumulate(data.begin(), data.end(), 0);
    }
};

class MaxReducer : public Reducer {
public:
    int combine(const std::vector<int>& data) const override {
        int best = data.front();
        for (int v : data) best = v > best ? v : best;
        return best;
    }
};

// Context delegates to the chosen strategy.
class Pipeline {
    std::unique_ptr<Reducer> reducer;
public:
    void setStrategy(std::unique_ptr<Reducer> r) { reducer = std::move(r); }
    int run(const std::vector<int>& data) const { return reducer->combine(data); }
};

int main() {
    std::vector<int> data{3, 7, 1, 9, 4};
    Pipeline pipeline;

    pipeline.setStrategy(std::make_unique<SumReducer>());
    std::cout << "sum = " << pipeline.run(data) << "\n";

    pipeline.setStrategy(std::make_unique<MaxReducer>());
    std::cout << "max = " << pipeline.run(data) << "\n";
    return 0;
}
